### PR TITLE
Feat/pin refactor

### PR DESCRIPTION
**Pull request by rosecers**
Refactoring the pinning infrastructure with respect to #45 

- [x] Do not call select({-1, -1}, guid) to indicate removal
- [ ] instead add specific functions for each possible action: 
    - [ ] addPinned 
    - [x] removePinned
    - [x] changeActivePinned. 
- [x] Then select always act on the current active
- [x] only move the marker/change the structure
- [x] Remove the starterGUID parameter for PropertyMap constructor to allow using the map without a structure viewer. This should be easy to do after the above changes
- [ ] Extract code dealing with selected markers (things like classList.toggle('chsp-active-structure-marker', false)) to a separate HTMLMarker class doing all of this. 
- [ ] On a related note, src/map/map.ts is getting quite large, so we should try to extract functionalities and potentially move it out of this file.

### CODE REVIEW

**Comment by rosecers** [comment]
@Luthaf can you elaborate on a few of these things? 

- What are the purpose of `addPinned` and `changeActivePinned` beyond `addWidget` and `set active`?
- We already only act on the active widget. What is different that is being requested?

File: src/utils/index.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 import {makeDraggable} from './draggable';
 import {DisplayMode, EnvironmentIndexer, Indexes} from './indexer';
-import {foreachSetting, HTMLSetting, SettingGroup, SettingModificationOrigin} from './settings';
+import {foreachSetting, HTMLSetting, SettingGroup, SettingModificationOrigin, Validator} from './settings';
```
Comment (Luthaf) [inline src/utils/index.ts:8]: Why is this required? The corresponding code does not exists on master

Comment (rosecers) [reply src/utils/index.ts:8]: On master, npm was throwing a compiler error for this not existing

Comment (Luthaf) [reply src/utils/index.ts:8]: Could to check again? Master is building fine without this: https://github.com/cosmo-epfl/chemiscope/runs/788274793

Comment (rosecers) [reply src/utils/index.ts]: Seems that the error wasn't recreated, so there might have been an artifact from creating the branch. I'll remove this line from the PR.

**Comment by rosecers** [comment]
> > What are the purpose of addPinned and changeActivePinned beyond addWidget and set active?
> 
> Mostly renaming to be consistent: `widget` does not make sense from the map point of view, hence the `pinned`; and having both `this.active = '...'` and `this.removeActive('...')` is strange, that's why I suggested to replace `set active()` with `changeActivePinned`.
> 
> > We already only act on the active widget. What is different that is being requested?
> 
> Removing the `activeGUID` parameter to 'onselect' and friends, which could be used to act on another widget.

So in other words -- onselect would always point to the active, but there would be a separate communicating function to change the active?

**Comment by Luthaf** [comment]
Exactly =)

**Comment by rosecers** [comment]
I think I'd like to leave the remaining issues for an additional PR, this is already getting somewhat big.

File: src/map/map.ts [hunk 2]
```diff
@@ -200,6 +200,8 @@ interface MarkerData {
 export class PropertiesMap {
     /** Callback fired when the plot is clicked and a new point is selected */
     public onselect: (indexes: Indexes, selectedGUID?: string) => void;
+    public onremove: (removedGUID: string) => void;
```
Comment (Luthaf) [inline src/map/map.ts:203]: This needs documentation

File: src/map/map.ts [hunk 3]
```diff
@@ -200,6 +200,8 @@ interface MarkerData {
 export class PropertiesMap {
     /** Callback fired when the plot is clicked and a new point is selected */
     public onselect: (indexes: Indexes, selectedGUID?: string) => void;
```
Comment (Luthaf) [inline src/map/map.ts:202]: Should the `selectedGUID` be removed here?

File: src/map/map.ts [hunk 4]
```diff
@@ -304,38 +298,40 @@ export class PropertiesMap {
     }
 
     /** Change the selected environment to the one with the given `indexes` */
-    public select(indexes: Indexes, selectedGUID: string = this._active) {
+    public select(indexes: Indexes, selectedGUID?: string) {
```
Comment (Luthaf) [inline src/map/map.ts:301]: This needs documentation for `selectedGUID`

File: src/map/map.ts [hunk 5]
```diff
@@ -346,6 +342,14 @@ export class PropertiesMap {
         this._settingsModal.remove();
     }
 
+    public get active() {
```
Comment (Luthaf) [inline src/map/map.ts:345]:
```suggestion
    public get active(): string {
```

This needs documentation

File: src/map/map.ts [hunk 6]
```diff
@@ -346,6 +342,14 @@ export class PropertiesMap {
...
+      if (this._selected.size > 0) {
+        if (this._active === undefined) {
+        this.active = this._selected.keys().next().value;
+        return this._selected.keys().next().value;
+        } else { return this._active; }
```
Comment (Luthaf) [inline src/map/map.ts:350]: minor nitpick: I prefer having the statements for the else case on a separate line, I feel it is easier to parse quickly when looking at the code

File: src/map/map.ts [hunk 7]
```diff
@@ -386,9 +392,40 @@ export class PropertiesMap {
...
+        this.activate(this._active);
     }
 
+    /**
+     * Removes a marker from the map.
+     *
+     * The parameter deleteFromGUIDs pertains to removing the guid from the guid
+     * list. This is *not* the same as the `force` parameter in
+     * `structure._removeWidget`. One may want to remove a marker when switching
+     * from 2D --> 3D but retain the information.
```
Comment (Luthaf) [inline src/map/map.ts:404]: I am not entirely happy with this: the function seems to do both hiding of HTML markers when switching from 2D to 3D; and marker removal. Could these two actions be separated easily?

File: src/map/map.ts [hunk 8]
```diff
@@ -1111,9 +1149,9 @@ export class PropertiesMap {
             // default to 0 (i.e. circles)
             return this._selectTrace<string | string[]>('circle', 'circle', trace);
         }
-
+        assert(this._selected.size > 0);
```
Comment (Luthaf) [inline src/map/map.ts:1152]: why is this assertion needed here?

File: src/structure/structure.ts [hunk 9]
```diff
@@ -83,7 +83,9 @@ interface WidgetGridData {
  */
 export class StructureViewer {
     /** Callback used when the user select an environment */
-    public onselect: (indexes: Indexes, selectedGUID?: string) => void;
+    public onselect: (indexes: Indexes) => void;
+    public onremove: (removedGUID: string) => void;
```
Comment (Luthaf) [inline src/structure/structure.ts:87]: needs documentation

File: src/structure/structure.ts [hunk 10]
```diff
@@ -169,7 +173,8 @@ export class StructureViewer {
         root.appendChild(this._root);
 
         this._setupGrid(1);
-        this._active = this._viewers.keys().next().value;
+        this.active = this._viewers.keys().next().value;
```
Comment (Luthaf) [inline src/structure/structure.ts:176]: Should we have an `assert(this._viewer.size > 0)` here?

Comment (Luthaf) [reply src/structure/structure.ts:176]: Or even extract this into a function `_firstPinned(): string`?

File: src/structure/structure.ts [hunk 11]
```diff
@@ -292,6 +296,34 @@ export class StructureViewer {
         }
     }
 
+    /*
+     * Removes a widget from the structure viewer grid.
+     * The parameter force pertains to removing the *only* widget in the grid,
+     * which should only be done when changing datasets.
```
Comment (Luthaf) [inline src/structure/structure.ts:302]: Since we are no longer changing datasets, should we remove this `force` parameter?

File: src/structure/structure.ts [hunk 12]
```diff
@@ -326,44 +358,40 @@ export class StructureViewer {
...
-        assert(data !== undefined);
-        let indexes;
-        if (this._indexer.mode === 'structure') {
-            indexes = this._indexer.from_structure_atom(data.current.structure);
-        } else {
-            const structure = data.current.structure;
-            const atom = data.current.atom;
-            indexes = this._indexer.from_structure_atom(structure, atom);
+        // choose *if* over assertion -- there's a weird error that I cannot pin-
+        // point where active is being passed an undefined.
```
Comment (Luthaf) [inline src/structure/structure.ts:363]: I'll try to give this a look, it might come from the constructor calling `this.active = '...'` before filling in `this.widgets`
